### PR TITLE
Generalizes keep effect induction code to combinatorial case

### PR DESCRIPTION
**Pull request by NishanthJKumar**
Joint PR with @wmcclinton 

Main things implemented:
* New test case for combinatorial machine example. This case illustrates two different situations: (1) where a combinatorial number of keep effect operators are needed for the `Configure` action, and (2) where only a subset of the full combinatorial number of keep effect operators are needed for the `Configure` action.
* Improvement to code that enables the backchaining approach to pass this above-mentioned test.

### CODE REVIEW

**Review by tomsilver — COMMENTED** [review]
this looks great, I don't have any complaints. I'll leave it to @ronuchit to give approval.

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py [hunk 1]
```diff
@@ -423,13 +424,21 @@ def _get_pnads_with_keep_effects(
         }
         if not keep_effects:
```
Comment (ronuchit) [inline src/nsrt_learning/strips_learning/gen_to_spec_learner.py:425]: we can take out this `if` statement now, right?

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py [hunk 2]
```diff
@@ -423,13 +424,21 @@ def _get_pnads_with_keep_effects(
...
             return set()
-        preconditions = pnad.op.preconditions | keep_effects
-        add_effects = pnad.op.add_effects | keep_effects
-        new_pnad_op = pnad.op.copy_with(name=f"{pnad.op.name}-KEEP",
-                                        preconditions=preconditions,
-                                        add_effects=add_effects)
-        new_pnad = PartialNSRTAndDatastore(new_pnad_op, [], pnad.option_spec)
-        return {new_pnad}
+        new_pnad_with_keep_effects = set()
```
Comment (ronuchit) [inline src/nsrt_learning/strips_learning/gen_to_spec_learner.py:427]: rename to `new_pnads_with_keep_effects`

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py [hunk 3]
```diff
@@ -423,13 +424,21 @@ def _get_pnads_with_keep_effects(
...
+        i = 0
+        for r in range(1, len(keep_effects) + 1):
+            for curr_keep_effects in itertools.combinations(keep_effects, r):
```
Comment (ronuchit) [inline src/nsrt_learning/strips_learning/gen_to_spec_learner.py:430]: rename to `keep_effects_subset` -- there's nothing really "current" about them

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py [hunk 4]
```diff
@@ -423,13 +424,21 @@ def _get_pnads_with_keep_effects(
...
-        preconditions = pnad.op.preconditions | keep_effects
-        add_effects = pnad.op.add_effects | keep_effects
-        new_pnad_op = pnad.op.copy_with(name=f"{pnad.op.name}-KEEP",
-                                        preconditions=preconditions,
-                                        add_effects=add_effects)
-        new_pnad = PartialNSRTAndDatastore(new_pnad_op, [], pnad.option_spec)
-        return {new_pnad}
+        new_pnad_with_keep_effects = set()
+        i = 0
+        for r in range(1, len(keep_effects) + 1):
```
Comment (ronuchit) [inline src/nsrt_learning/strips_learning/gen_to_spec_learner.py:429]: add a comment before this loop describing what the whole deal is

**Review by ronuchit — APPROVED** [review]
looks good! just left a few minor comments

**Comment by NishanthJKumar** [comment]
Supercloud results:

```
Git commit hashes seen in results/:
4a7498cdf5444822d9cf06f3469073c556750b5d


AGGREGATED DATA OVER SEEDS:
                               NUM_TRAIN_TASKS    NUM_SOLVED AVG_NUM_PREDS AVG_TEST_TIME  AVG_NODES_CREATED    LEARNING_TIME  NUM_SEEDS
EXPERIMENT_ID                                                                                                                          
rnt_backchain                     50.00 (0.00)  50.00 (0.00)   3.00 (0.00)   0.05 (0.00)       88.50 (2.35)     15.14 (4.63)         10
rnt_no_sidepreds                  50.00 (0.00)   4.89 (4.70)   3.00 (0.00)   3.51 (3.57)  1484.92 (1365.34)    90.21 (92.57)          9
rnt_oracle                         0.00 (0.00)  50.00 (0.00)   3.00 (0.00)   0.02 (0.00)       88.50 (2.35)      0.00 (0.00)         10
rnt_painting_backchain            50.00 (0.00)  49.50 (0.67)  18.00 (0.00)   0.78 (0.11)   1836.69 (446.50)    101.61 (8.67)         10
rnt_painting_oracle                0.00 (0.00)  49.40 (1.02)  18.00 (0.00)   1.36 (0.15)  6874.46 (2075.80)      0.00 (0.00)         10
rnt_pred_error                    50.00 (0.00)  45.60 (4.41)   3.00 (0.00)   0.50 (0.35)       91.18 (2.39)    52.79 (74.23)         10
rnt_single_option_backchain       50.00 (0.00)  50.00 (0.00)   3.00 (0.00)   0.04 (0.01)       88.50 (2.35)     28.43 (5.68)         10
rnt_single_option_no_sidepreds    50.00 (0.00)   4.67 (4.55)   3.00 (0.00)   3.25 (3.40)  1438.15 (1308.87)  147.42 (180.60)          9
rnt_single_option_oracle           0.00 (0.00)  50.00 (0.00)   3.00 (0.00)   0.02 (0.00)       88.50 (2.35)      0.00 (0.00)         10
rnt_single_option_pred_error      50.00 (0.00)  43.40 (2.76)   3.00 (0.00)   0.24 (0.16)      67.59 (26.88)   88.61 (139.84)         10

```

**Pull request merged**